### PR TITLE
Support for multitenancy

### PR DESCRIPTION
**Pull request by tobiasmcnulty**
- Add a domain to the SiteMode model and use that to discover a SiteMode based on the request, rather than supporting only one SiteMode.
- Add FKs to all the appropriate models and filters on the appropriate queries so questions/voters, etc. are restricted to a single site.
- Still support username logins for backwards compatibility, but prefer email address logins and restrict those to the current site, in case the same user signs up on multiple sites.

To do:
- [ ] Remove the username from the registration field and auto-generate a unique username based on the email address
- [ ] Add tests for 2+ simultaneous sites (e.g., questions show up only on their respective sites)

### CODE REVIEW

File: opendebates/views.py [hunk 1]
```diff
@@ -98,9 +97,9 @@ def list_ideas(request):
 @rendered_with("opendebates/list_ideas.html")
 def list_category(request, cat_id):
     category = get_object_or_404(Category, id=cat_id)
```
Comment (ejucovy) [inline opendebates/views.py]: `get_object_or_404(Category, id=cat_id, site_mode=request.site_mode)`?

File: opendebates/views.py [hunk 2]
```diff
@@ -171,6 +174,7 @@ def vote(request, id):
 
     if request.method == "GET":
         two_other_approved_ideas = list(Submission.objects.filter(
+            category__site_mode=request.site_mode,
```
Comment (ejucovy) [inline opendebates/views.py]: Is this necessary?  We already have `category=idea.category` in the filter so hopefully it's redundant.

File: opendebates/models.py [hunk 3]
```diff
@@ -32,6 +33,11 @@ class Meta:
 
 
 class SiteMode(CachingMixin, models.Model):
+    THEME_CHOICES = [(theme, theme) for theme in settings.SITE_THEMES]
+
+    domain = models.CharField(_('domain name'), unique=True, max_length=100)
+    theme = models.CharField(max_length=255, choices=THEME_CHOICES)
```
Comment (ejucovy) [inline opendebates/models.py]: Is this+SITE_THEMES really doing anything for us anymore?  I _think_ the only thing that `theme` is providing now is an automatic css inclusion from the codebase.  Now that we have `mode. inline_css` maybe we should just get rid of that altogether instead of having this awkward dependency on code state?

(But maybe that ought to be a separate PR/thread.)

**Review by ejucovy — COMMENTED** [review]
A few tiny comments inline.  Also:
- Should ENABLE_USER_DISPLAY_NAME and ENABLE_USER_PHONE_NUMBER go onto sitemode in this case?  I guess that can happen later.
- Do we need to worry about `settings.SITE_DOMAIN` and SITE_DOMAIN_WITH_PROTOCOL?  There's various (unchanged) bits of code that use those to generate URLs in emails and share links, etc -- e.g. see `opendebates.models:Submission.really_absolute_url`
- Should `opendebates_emails.models:EmailTemplate` get attached to a sitemode?  I think it might be necessary but not sure.  Would also require some changes to the code that triggers those.
- Might actually even make sense to move MIXPANEL_KEY, OPTIMIZELY_KEY, NORECAPTCHA\* from settings onto SiteMode depending on whether those services' api keys are attached to a domain.  (I _think_ they are but would need to double check.)

**Comment by jbradberry** [comment]
There are values in opendebates/site_defaults.py that will probably need to be dealt with differently under this scheme.

File: opendebates/views.py [hunk 4]
```diff
@@ -423,11 +441,11 @@ def report(request, id):
 @allow_http("GET", "POST")
 @login_required
 def merge(request, id):
-    if not allow_voting_and_submitting_questions() and not request.user.is_staff:
+    if not request.site_mode.allow_voting_and_submitting_questions and not request.user.is_staff:
         raise Http404
 
     idea = get_object_or_404(Submission, pk=id)
```
Comment (jbradberry) [inline opendebates/views.py]: `get_object_or_404(Submission, pk=id, site_mode=request.site_mode)`?  Presumably we're not sharing questions between sites.

**Comment by jbradberry** [comment]
moderator_views.remove has a `get_object_or_404` that should probably be updated.

File: opendebates/views.py (hunk 4 above)
Comment (jbradberry) [inline opendebates/views.py]: er, `category__site_mode=...`

File: opendebates/views.py [hunk 5]
```diff
@@ -399,11 +417,11 @@ def list_candidates(request):
 @allow_http("GET", "POST")
 @login_required
 def report(request, id):
-    if not allow_voting_and_submitting_questions() and not request.user.is_staff:
+    if not request.site_mode.allow_voting_and_submitting_questions and not request.user.is_staff:
         raise Http404
 
     idea = get_object_or_404(Submission, pk=id)
```
Comment (jbradberry) [inline opendebates/views.py]: `get_object_or_404(Submission, pk=id, category__site_mode=request.site_mode)`

File: opendebates/views.py (hunk 4 above)
Comment (jbradberry) [inline opendebates/views.py]: and here.

**Comment by jbradberry** [comment]
For the most part, this codebase is using Django's built-in views for auth and registration.  Django uses `get_current_site` to determine the domain and site name for rendering those views (or rendering the emails sent), which I think is not adjusted by this scheme.

File: opendebates/middleware.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    Gets or creates a SiteMode for the request, based on the hostname.
+    """
+
+    def process_request(self, request):
+        request.site_mode = get_site_mode(request)
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        if 'prefix' in view_kwargs:
+            view_kwargs.pop('prefix')
+            if not getattr(request, 'site_mode', None):
```
Comment (tobiasmcnulty) [inline opendebates/middleware.py:17]: `if getattr(...) is None:` would read more clearly to me here.

File: opendebates/urls.py [hunk 7]
```diff
@@ -2,39 +2,46 @@
 from django.contrib import admin
 
 urlpatterns = [
-    url('^$', 'opendebates.views.list_ideas', name="list_ideas"),
+    url(r'^admin/', include(admin.site.urls)),
```
Comment (tobiasmcnulty) [inline opendebates/urls.py:5]: To avoid repeating `(?P<prefix>[-\w]+)/` & help pare down the diff, it might be nice to keep this file as-is but move the URLs that _don't_ take a prefix to a separate urls file, and include this URLs file with the prefix there.

File: opendebates/templates/base.html [hunk 8]
```diff
@@ -65,14 +65,14 @@
     </div>
     <div class="login-container{% if request.user.is_authenticated %} login-container-authenticated{% endif %}">
       {% if request.user.is_authenticated %}
-        <a href="{# url 'user_profile' #}">
+        <a href="{# url 'user_profile' SITE_MODE.prefix #}">
```
Comment (tobiasmcnulty) [inline opendebates/templates/base.html:68]: Might be worth checking what the intended behavior is here... a link with no `href` seems odd (why not just remove the link altogether?)...

**Comment by tobiasmcnulty** [comment]
- Most recent commit looks good
- Did a few smoke tests of this locally with the `prod01` data; with the exception of the `/recent/` JS needing the correct URL, didn't see any issues (confirmed I can migrate, add a prefix for NH senate, see the old data, add a new site, and submit/vote on that question)
- Added some minor display enhancements for editing site modes via the admin
- Updated the Site domains on the live `prod01` and `staging01` servers to correct values, so we should be able to get in after this is deployed (for some reason these were set to example.com, which also broke password reset emails)

Pending a fix to the JS `/recent/` issue, consider this a :shipit: from me to merge & deploy to staging for further testing.
